pushers/channel.go

```diff
@@ -15,81 +14,58 @@ var log = logging.MustGetLogger("honeytrap:channels")
+// to filter out the desired messages allowed for all registered backends.
+func (mc FilterChannel) Send(msgs ...message.Event) {
+	mc.Backend.Send(mc.Filter.Filter(msgs...)...)
 }
 
-// UnmarshalConfig attempts to unmarshal the provided value into the target
-// FilteringChannel.
-func (mc *FilteringChannel) UnmarshalConfig(m interface{}) error {
-	conf, ok := m.(config.ChannelConfig)
-	if !ok {
-		return errors.New("Expected to receive a ChannelConfig type")
-	}
+// MakeFilter returns a slice of Channels which match the giving criterias
+// associated with the provided config.ChannelConfig.
+func MakeFilter(config *config.Config, conf config.ChannelConfig) ([]Channel, error) {
+	var filters FilterGroup
+	filters = append(filters, NewRegExpFilter(CategoryFilterFunc, MakeMatchers(conf.Categories...)...))
+	filters = append(filters, NewRegExpFilter(SensorFilterFunc, MakeMatchers(conf.Sensors...)...))
+	filters = append(filters, NewRegExpFilter(EventFilterFunc, MakeMatchers(conf.Events...)...))
+
+	var channels []Channel
 
 	// Generate all filters for the channel's backends
 	for _, backend := range conf.Backends {
 
 		// Retrieve backend configuration.
-		backendPrimitive, ok := mc.config.Backends[backend]
+		backendPrimitive, ok := config.Backends[backend]
 		if !ok {
-			return fmt.Errorf("Application has no backend named %q", backend)
+			return nil, fmt.Errorf("Application has no backend named %q", backend)
 		}
 
 		var item = struct {
 			Backend string `toml:"backend"`
 		}{}
 
-		if err := mc.config.PrimitiveDecode(backendPrimitive, &item); err != nil {
-			return err
+		if err := config.PrimitiveDecode(backendPrimitive, &item); err != nil {
+			return nil, err
 		}
 
 		// Attempt to create backend channel for master with the giving
 		// channel's name and config toml.Primitive.
-		newBackend, err := NewBackend(item.Backend, mc.config.MetaData, backendPrimitive)
+		base, err := NewBackend(item.Backend, config.MetaData, backendPrimitive)
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		mc.backends = append(mc.backends, newBackend)
-	}
-
-	mc.filters = append(mc.filters, NewRegExpFilter(CategoryFilterFunc, MakeMatchers(conf.Categories...)...))
-	mc.filters = append(mc.filters, NewRegExpFilter(SensorFilterFunc, MakeMatchers(conf.Sensors...)...))
-	mc.filters = append(mc.filters, NewRegExpFilter(EventFilterFunc, MakeMatchers(conf.Events...)...))
-
-	return nil
-}
-
-// Send delivers the slice of PushMessages and using the internal filters
-// to filter out the desired messages allowed for all registered backends.
-func (mc *FilteringChannel) Send(msgs ...message.Event) {
-
-	// filter messages with all filters.
-	for _, filter := range mc.filters {
-		msgs = filter.Filter(msgs...)
-	}
-
-	// If no message passes our filtering conditions then we have
-	// nothing else to do here.
-	if len(msgs) == 0 {
-		return
+		channels = append(channels, FilterChannel{
```

Comment:
What about filterChannel as struct, and FilterChannel method which will return a new FilterChannel{channel, filters}?